File: api/src/main/java/me/kvalbrus/multibans/api/punishment/Cancelable.java
package me.kvalbrus.multibans.api.punishment;

import me.kvalbrus.multibans.api.punishment.creator.PunishmentCreator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface Cancelable {

    void deactivate(@Nullable PunishmentCreator cancellationCreator, long cancellationDate, @Nullable String cancellationReason);

    @Nullable
    PunishmentCreator getCancellationCreator();

    long getCancellationDate();

    @Nullable
    String getCancellationReason();

    boolean isCancelled();

    void setCancelled(boolean cancelled);

    void setCancellationCreator(@Nullable PunishmentCreator cancellationCreator);

    void setCancellationDate(long cancellationDate);

    void setCancellationReason(@NotNull String cancellationReason);
}

File: api/src/main/kotlin/me/kvalbrus/multibans/api/punishment/Cancelable.kt
package me.kvalbrus.multibans.api.punishment

import me.kvalbrus.multibans.api.punishment.action.Action
import me.kvalbrus.multibans.api.punishment.action.ActivationAction
import me.kvalbrus.multibans.api.punishment.action.DeactivationAction
import me.kvalbrus.multibans.api.punishment.executor.PunishmentExecutor

interface Cancelable {

    fun activate(action: ActivationAction) : Boolean
    fun activate(executor: PunishmentExecutor, date: Long, reason: String) : Boolean
    fun deactivate(action: DeactivationAction) : Boolean
    fun deactivate(executor: PunishmentExecutor, date: Long, reason: String) : Boolean

    val activations: List<Action>
    val deactivations: List<Action>

    fun lastActivator() : PunishmentExecutor
    fun lastDeactivator() : PunishmentExecutor?

    val cancelled : Boolean

//    var cancellationCreator: PunishmentCreator?
//    var cancellationDate: Long
//    var cancellationReason: String
//    var cancelled: Boolean
}